staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,587 @@
+				value:      m,
+				schema:     schema,
+				propSchema: func(key string) *structuralschema.Structural { schema := schema.Properties[key]; return &schema },
+			}
+		}
+		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Structural != nil {
+			return &unstructuredMap{
+				value:      m,
+				schema:     schema,
+				propSchema: func(key string) *structuralschema.Structural { return schema.AdditionalProperties.Structural },
+			}
+		}
+		return types.NewErr("invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema")
+	}
+	if schema.Type == "array" {
+		if schema.Items == nil {
+			return types.NewErr("invalid array type, expected Items with a non-empty Schema")
+		}
+		l, ok := unstructured.([]interface{})
+		if !ok {
+			return types.NewErr("invalid data, expected []interface{} to match the provided schema with type=array")
+		}
+		typedList := unstructuredList{elements: l, itemsSchema: schema.Items}
+		listType := schema.XListType
+		if listType != nil {
+			switch *listType {
+			case "map":
+				mapKeys := schema.Extensions.XListMapKeys
+				return &unstructuredMapList{unstructuredList: typedList, escapedKeyProps: escapeKeyProps(mapKeys)}
+			case "set":
+				return &unstructuredSetList{unstructuredList: typedList}
+			case "atomic":
+				return &typedList
+			default:
+				return types.NewErr("invalid x-kubernetes-list-type, expected 'map', 'set' or 'atomic' but got %s", *listType)
+			}
+		}
+		return &typedList
+	}
+
+	if schema.Type == "string" {
+		str, ok := unstructured.(string)
+		if !ok {
+			return types.NewErr("invalid data, expected string, got %T", unstructured)
+		}
+		if schema.ValueValidation != nil {
+			switch schema.ValueValidation.Format {
+			case "duration":
+				d, err := strfmt.ParseDuration(str)
+				if err != nil {
+					return types.NewErr("Invalid duration %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(d)
+			case "date":
+				d, err := time.Parse(strfmt.RFC3339FullDate, str) // strfmt uses this format for OpenAPIv3 value validation
+				if err != nil {
+					return types.NewErr("Invalid date formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(d)
+			case "date-time":
+				d, err := strfmt.ParseDateTime(str)
+				if err != nil {
+					return types.NewErr("Invalid date-time formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(time.Time(d))
+			case "byte":
+				base64 := strfmt.Base64{}
+				err := base64.UnmarshalText([]byte(str))
+				if err != nil {
+					return types.NewErr("Invalid byte formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue([]byte(base64))
+			}
+		}
+		return types.DefaultTypeAdapter.NativeToValue(unstructured)
+	}
+	if schema.Type == "number" || schema.Type == "integer" || schema.Type == "boolean" {
+		return types.DefaultTypeAdapter.NativeToValue(unstructured)
```

Comment:
should we make sure the type of unstructured matches? if openapi is guarding things for us, boolean and integer might be safe already, but I'm not sure what happens with number... openapi might tolerate an int for that... is CEL expecting a float64?

The way we decode, `1` will decode as an int64, and `1.0` or `1.5` will decode as a float64... I suspect you could encounter either an int64 or a float64 for a type=number field. Especially until cel has heterogenous equality, we should probably make sure fields of type number consistently get a float64

Reply:
Yes, I've made all type checks strict. I allow all int and float types that `DefaultTypeAdapter.NativeToValue` supports, but unstructured only uses a subset of them, so we could tighten it down... I don't know if it really matters that we are more accepting than the data we know we'll actually be processing.

Reply:
`float` representations of whole numbers (e.g. `1.0`, `0.0`) can convert to int representations (e.g. `1`, `0`) in yaml → json translation, and then get parsed as int64s.

For `type=number`, we need to allow both float64 and int64 and convert int64 to float64 so it works with CEL expressions expecting floats.

This was specifically visible in this test case:
* https://gist.github.com/liggitt/5bb0ece637e560ec906ee6a0786eb3bd#file-cel-type-mismatch-yaml-L61
* https://gist.github.com/liggitt/5bb0ece637e560ec906ee6a0786eb3bd#file-results-txt-L3-L4

Reply:
Fixed. This required me skip the DefaultTypeAdapter.NativeToValue call and pick the types we map to directly, which in hindsight it simpler and clearer.